File: rel-tree/src/main/java/hue/captains/singapura/js/homing/reltree/RelTreeChannelModule.java
package hue.captains.singapura.js.homing.reltree;

import hue.captains.singapura.js.homing.core.DomModule;
import hue.captains.singapura.js.homing.core.Exportable;
import hue.captains.singapura.js.homing.core.ExportsOf;
import hue.captains.singapura.js.homing.core.ImportsFor;
import hue.captains.singapura.js.homing.core.ModuleImports;

import java.util.List;
import hue.captains.singapura.js.homing.relchannel.RelChannelModule;
import hue.captains.singapura.js.homing.relgrid.protocol.RelGridProtocolModule;

/**
 * RFC 0050 · Episode 3-ext1 — {@code RelTreeChannel}: the tree's customers of the
 * ask channel, over the core in {@code rel-channel}. Unfold and fold are
 * questions answered with a {@code RelTreeView} or nothing; the cursor and an
 * activation are notifications. The payloads are the protocol's generated
 * value objects, which with the core is all this imports.
 */
public record RelTreeChannelModule() implements DomModule<RelTreeChannelModule> {
    public record RelTreeChannel() implements Exportable._Constant<RelTreeChannelModule> {}
    public static final RelTreeChannelModule INSTANCE = new RelTreeChannelModule();
    @Override public ImportsFor<RelTreeChannelModule> imports() {
        return ImportsFor.<RelTreeChannelModule>builder()
                .add(new ModuleImports<>(List.of(new RelChannelModule.RelChannel()), RelChannelModule.INSTANCE))
                .add(new ModuleImports<>(
                        List.of(new RelGridProtocolModule.RelTreeView(),
                                new RelGridProtocolModule.RelTreeUnfold(),
                                new RelGridProtocolModule.RelTreeFold(),
                                new RelGridProtocolModule.RelTreeCursorChanged(),
                                new RelGridProtocolModule.RelTreeActivated()),
                        RelGridProtocolModule.INSTANCE))
                .build();
    }
    @Override public ExportsOf<RelTreeChannelModule> exports() {
        return new ExportsOf<>(INSTANCE, List.of(new RelTreeChannel()));
    }
}

File: rel-tree/src/main/resources/homing/js/hue/captains/singapura/js/homing/reltree/RelTreeChannelModule.js
// =============================================================================
// RelTreeChannelModule — RFC 0050 · Episode 3-ext1's ASK CHANNEL, the tree's
// side: the tree's CUSTOMERS of the core in rel-channel. Two questions —
// unfold and fold — each answered with a View to present whole, or nothing;
// two notifications — the cursor moved, a node was activated — handed over
// and never waited on. The waiting, the mask, the one-at-a-time are the
// core's, and the layout is its surface.
//
// FOLD AND UNFOLD ARE BOTH QUESTIONS, answered with the whole View. Not the
// children of one node: the tree holds no structure to insert them into,
// and the fold state is the relation's — a domain that folds a node may
// free what was under it, and a domain that refuses answers nothing. LAZY
// BY DEFAULT: an unfold is the only way children are ever asked for.
//
// The payloads are the protocol's generated value objects. The tree mints
// them and never reads them back; an answer is presented exactly as given,
// or not at all.
//
//   new RelTreeChannel({ ask, layout, onView })
//   unfold(key) / fold(key)         the questions; true when asked
//   cursorChanged(key) / activated(key)   the notifications
//   has() / isPending() / destroy()
// =============================================================================

class RelTreeChannel {

    constructor(opts) {
        this._core = new RelChannel({ ask: opts.ask || null, surface: opts.layout, tag: "[RelTree]" });
        this._onView = opts.onView;               // (places) — the facade presents; it may throw, and the core records it
    }

    has()       { return this._core.has(); }
    isPending() { return this._core.isPending(); }
    destroy()   { this._core.destroy(); }

    // ── the questions ──────────────────────────────────────────────────────

    unfold(key) {
        var self = this;
        return this._core.ask(new RelTreeUnfold(key), function (answer) { self._applyView(answer, "unfold"); });
    }

    fold(key) {
        var self = this;
        return this._core.ask(new RelTreeFold(key), function (answer) { self._applyView(answer, "fold"); });
    }

    /** A View is presented, exactly as given; absence leaves the rows as they are; a malformed one is refused whole. */
    _applyView(answer, kind) {
        if (answer == null) return;
        if (!(answer instanceof RelTreeView)) {
            console.error("[RelTree] a " + kind + " was answered with something that is not a tree View:", answer);
            return;
        }
        this._onView(answer.places, kind);
    }

    // ── the notifications ──────────────────────────────────────────────────

    cursorChanged(key) { if (this._core.has()) this._core.notify(new RelTreeCursorChanged(key)); }
    activated(key)     { if (this._core.has()) this._core.notify(new RelTreeActivated(key)); }
}
